lib/internal/url.js

```diff
@@ -605,11 +606,34 @@ class URL {
     ctx.password = password;
     ctx.port = port;
     ctx.hash = hash;
-    if (!this[searchParams]) { // Invoked from URL constructor
-      this[searchParams] = new URLSearchParams();
+    if (this[searchParams]) {
+      // Update `kDirty` property to recalculate searchParams on access.
+      // This is done to reduce the overhead of initializing the URL.
+      this[searchParams][kDirty] = true;
+    }
+  };
+
+  #onSearchUpdate = (href, origin, protocol, hostname, pathname,
+                     search, username, password, port, hash) => {
+    const ctx = this[context];
+    ctx.href = href;
+    ctx.origin = origin;
+    ctx.protocol = protocol;
+    ctx.hostname = hostname;
+    ctx.pathname = pathname;
+    ctx.search = search;
+    ctx.username = username;
+    ctx.password = password;
+    ctx.port = port;
+    ctx.hash = hash;
+
+    if (this[searchParams] == null) {
+      this[searchParams] = new URLSearchParams(this[context].search);
       this[searchParams][context] = this;
```

Comment:
Why do we need to create a `URLSearchParams` here? If it's null, that can only mean that no one has asked to access it. Then later in the `searchParams` getter it'll be populated anyway.